docs/user/exercises/programming-exercise-features.inc

```diff
@@ -87,3 +90,46 @@ Instructors can still use those templates to generate programming exercises and
 
 We encourage instructors to contribute improvements to the existing ``templates`` or to provide new templates. Please contact Stephan Krusche and/or create Pull Requests in the GitHub repository.
 
+.. _DejaGnu:
+
+Java DejaGnu: Blackbox Testing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The classical testing frameworks, like JUnit for example, allows writing white box tests.
+This enforces many assumptions about the structure of the source code under test,
+like the class and methods names and even the exact method signatures.
+This implies for a Programming Exercise you have to specify all the structural aspect
+if your test suite should be able to run on the student submissions.
+That is maybe okay or even desired for a beginner course,
+but for advanced courses this is a downside,
+because without making their own decision about the structure of their source code,
+the students cannot collect experience in this important aspect of programming through your exercises.
+
+To be able to include the structural aspect of programming in your Programming Exercises,
+without losing the ability to test submissions automatically, you can use DejaGnu.
+It allows you to write blackbox tests for command line interfaces of applications.
+The tests are written in Expect Script, an extension to the Tcl scripting language.
+Expect is a traditional Unix utility for automatic interactions with programs that expose a text terminal
+interface in a robust way.
+The test script starts the program to test as own process (possibly several times) and interactively controls
+it by textual inputs via pipelining.
```

Comment:
```suggestion
it by textual inputs via the standard input.
```
Pipelining definitely does not make sense here. I guess you mean ‘piping the text from the tool into the program’. I assume calling it standard input should be less confusing to others unfamiliar with the tool.